qiskit/circuit/parameter.py

```diff
@@ -90,6 +90,24 @@ def __init__(self, name: str):
             symbol = symengine.Symbol(name)
         super().__init__(symbol_map={self: symbol}, expr=symbol)
 
+    def assign(self, parameter, value):
+        if parameter != self:
+            # Corresponds to superclass calls to `subs` and `bind` that would implicitly set
+            # `allow_unknown_parameters=False`.
+            raise CircuitError(
+                f"Cannot bind Parameters ({[str(parameter)]}) not present in expression."
+            )
+        if isinstance(value, ParameterExpression):
+            # This is the `super().subs` case.
+            return value
+        # This is the `super().bind` case, where we're required to return a `ParameterExpression`,
+        # so we need to lift the given value to a symbolic expression.
```

Comment:
Why do we need to lift the value to a symbolic expression? In the final circuit keeping only the float would be even better than having the parameter expression, but I expect there's some type requirement I'm missing...

Reply:
In this PR at least, because that would be a different logical change.  I think there's a few issues discussing your topic elsewhere on the Terra repo, though.